Add a command-line interface and README

freshstrangemusic opened the pull request.

brennie committed:
Add a command line interface with structopt

This crate is split up into a library and two binaries:

- fxrunner: the binary responsible for running Firefox on the reference
  hardware;
- fxrecorder: the binary responsible for doing frame capture; and
- fxrecord: the library with common code for both fxrunner and
  fxrecorder.

The configuration for both fxrunner and fxrecorder is parsed from a TOML
file (configurable via the `--config` command-line option). For ease of
development, the fxrunner and fxrecorder configuration are both able to
be included in a single config file (`fxrecord.toml`) using sections. An
example configurationf ile (`fxrecord.example.toml`) is provided.

brennie committed:
Add a README

brennie committed:
Use slog for logging

slog et al. are robust logging libraries providing structured logging.

hawkinsw commented on README.md:
```diff
@@ -0,0 +1,96 @@
+`fxrecord` is a tool for capturing and analyzing recordings of Firefox
+desktop startup. It consists of two parts: `fxrecorder` and `fxrunner`. The
+server does the capturing and analysis. It runs on a separate device from the
+client. The client runs on the machine that is to be recorded and is
```
Here, and throughout the document, it seems like 'server' and 'client' are used somewhat interchangeably with 'recorder' and 'runner', respectively. Is this correct? If so, can you clarify that?

freshstrangemusic replied:
Oops yeah I renamed it from client/server to recorder/runner and I guess I missed some.

hawkinsw commented on src/bin/fxrecorder/main.rs:
```diff
@@ -0,0 +1,49 @@
+    println!("options = {:#?}", options);
+
+    let config: Config =
+        read_config(&options.config_path, "fxrecorder").context(InvalidConfig {})?;
```
Could we comment this somehow to be more descriptive of where `.context()` comes from? In other words, point to the fact that the `Snafu::ResultExt` trait is implemented for all `Result<>`s?

freshstrangemusic replied:
So .context() is going to show up in a *lot* of places. I don't think it makes sense to have a comment everywhere we wrap an error.

A modern editor will have goto definition support for `.context()` and `FooExt` traits are pretty common, especially `ResultExt`-style ones.

hawkinsw commented on src/lib/lib.rs:
```diff
@@ -0,0 +1,80 @@
+    for<'de> T: Deserialize<'de>,
+    P: AsRef<Path>,
+{
+    let path = path.as_ref();
```
Can you be explicit about why you do this rather than `&path` everywhere?

freshstrangemusic replied:
This allows you to pas anything that can be derefed into a Path, e.g., Path, PathBuf, OsStr, String, &str, etc. It is quite a common pattern in the standard library, e.g. [`std::path::Path::strip_prefix`](https://doc.rust-lang.org/std/path/struct.Path.html#method.strip_prefix), [`std:path::Path::starts_with`](https://doc.rust-lang.org/std/path/struct.Path.html#method.starts_with), [`std::path::Path::ends_with`](https://doc.rust-lang.org/std/path/struct.Path.html#method.ends_with)

hawkinsw replied:
*I* know that, but I was hoping you could document it for "future" me.

hawkinsw commented on src/bin/fxrunner/main.rs:
```diff
@@ -0,0 +1,50 @@
+    let log = build_logger();
+
+    if let Err(e) = fxrunner(log.clone(), options) {
+        error!(log, "unexpected error"; "error" => %e);
```
This is a little incongruous for the end user because the output has a very specific error message (parsing error, I/O error, etc) and then it's prefixed by 'unexpected error'. Perhaps you could just print the error and not the prefix?

Same would apply to fxrecorder, too.

freshstrangemusic replied:
the unexpected error bit is because these errors should not happen, i.e., these are the exceptional cases if they've bubbled up all the way up to main.

Most errors that occur won't make it this far.

hawkinsw replied:
I think that they will. I ran the tool and saw this very output in an unexceptional, exceptional situation: a field in the input that is not supported by the containing section.

freshstrangemusic replied:
Also part of this is we are using structured logging, where we can set Key-Value pairs. We might end up switching to a logging format that, eg uses JSON and then the error format would look like

```json
{
     "event": "unexpected error",
     "error": "could not open config file"
}
```

In this case, if we put the error message in the event field, we wouldn't be able to grep through the logs for all errors since there would be no error key.

Also, not being able to parse the config file is exceptional. If we cant interpret it, then we cant run.

hawkinsw replied:
All I am saying is that if we have code that handles the error and formats it, then it's not unexpected. I see unexpected as an error that the code was not designed to handle. So, something like this:

```
Oct 07 17:29:16.818 ERRO unexpected error, error: Could not parse config file `fxrecord.toml': missing field `host`
```

does not make sense. We have enough "intelligence" in the code to format the error properly, and report it succinctly and informatively. It shouldn't be labeled unexpected. 

Yes, I am bikeshedding. Ultimately it's your choice.

freshstrangemusic replied:
These logs are not intended for human consumption. People aren't going to be running this command. Its going to be triggered by try jobs and the results will be reported there.

hawkinsw commented on src/bin/fxrunner/main.rs:
```diff
@@ -0,0 +1,50 @@
+use snafu::{ResultExt, Snafu};
+use structopt::StructOpt;
+
+use crate::config::Config;
```
This is probably obvious to Rustaceans, but it might be good to say that this points to the mod's `Config` and *not* [this](https://docs.rs/config/0.9.3/config/struct.Config.html).

Same would apply to fxrecorder, too.

freshstrangemusic replied:
In Rust 2018, things prefixed by `crate` are scoped to the local crate. If it was from the `config` crate, it would be `config::Config` (and would be grouped with other non-crate imports above).

hawkinsw replied:
Again, *I* know that but I was hoping we could make it explicit for "future" me.

brennie commented:
Im going to update this PR to remove snafu. I've been using it in further patches and am not happy with using for handling errors, esp. namespace pollution and name conflicts.

brennie commented:
@hawkinsw I ripped out snafu and all the magic that went with it. These patches should be clearer wrt error types, albeit a bit more verbose.

brennie committed:
Refactor common logic from main()s

Both fxrunner's main and fxrecorder's main did basically the same thing
(set up a logger, read options from the command-line, and then call a
function with the logger and options). In addition, both called
functions also read in a config file, where the path was given by the
command-line options. They also redefined the same Error enum.

This has been abstracted into the `fxrecord::run` function, which does
all of the above. We are now returning a `Result<(), Box<dyn Error>>`
because at the top level all errors are unexpected. Additionally, A
`CommonOptions` trait has been introduced so that `fxrecord::run` can
load the config to pass to the runnable.

smacleod commented on src/bin/fxrecorder/config.rs:
```diff
@@ -0,0 +1,67 @@
+    /// The frame rate to capture.
+    ///
+    /// This corresponds to the `-framerate` argument to `ffmpeg`.
+    pub frame_rate: u8,
```
While 255fps is probably enough, I wouldn't be opposed to bumping this higher... heh.

freshstrangemusic replied:
:)

smacleod reviewed (approved):
LGTM, just one comment to think about or ignore.... Need that super super high frame rate recording ;)
